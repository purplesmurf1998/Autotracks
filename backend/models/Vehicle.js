const mongoose = require('mongoose');
const Comment = require('./Comment');
const History = require('./History');
const Event = require('./Event');
const VehicleSale = require('./VehicleSale');

const VehicleSchema = new mongoose.Schema({
  dealership: {
    type: mongoose.Schema.ObjectId,
    ref: 'Dealership',
    required: [true, 'Vehicle model must be associated to a dealership']
  },
  vin: {
    type: String,
    required: [true, 'Vehicle must have a \'Vehicle Identification Number\' (VIN)'],
    unique: true,
    minlength: [11, 'VIN must be 17 characters long, or at least 11 characters for vehicles manufactured before 1981'],
    maxlength: [17, 'VIN must be 17 characters long, or at least 11 characters for vehicles manufactured before 1981']
  },
  missing: {
    type: Boolean,
    default: false
  },
  delivered: {
    type: Boolean,
    default: false
  },
  date_delivered: Date,
  sale: {
    type: mongoose.Schema.ObjectId,
    ref: 'VehicleSale',
    default: null
  },
  date_added: {
    type: Date,
    default: Date.now
  },
  on_road_since: Date,
  properties: Object,
})

// cascade delete comments, history, vehiclesale
VehicleSchema.pre('remove', async function (next) {
  await Comment.deleteMany({ vehicle: this._id });
  await History.deleteMany({ vehicle: this._id });
  await VehicleSale.deleteMany({ vehicle: this._id });
  await Event.deleteMany({ vehicle: this._id });
});

module.exports = mongoose.model('Vehicle', VehicleSchema);