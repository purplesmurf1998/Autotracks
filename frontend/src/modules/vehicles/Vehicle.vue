<template>
  <div>
    <CRow>
      <CCol>
        <vehicle-details :vehicle="vehicle" :setNewVehicle="setNewVehicle"/>
      </CCol>
      <CCol>
        <vehicle-location :vehicle="vehicle" />
      </CCol>
    </CRow>
    <vehicle-properties :vehicle="vehicle" :setNewVehicle="setNewVehicle"/>
  </div>
</template>

<script>
import VehicleDetails from "./VehicleDetails.vue";
import VehicleLocation from "./VehicleLocation.vue";
import VehicleProperties from "./VehicleProperties.vue";

export default {
  name: 'Vehicle',
  props: ["vehicle", "setNewVehicle"],
  components: {
    "vehicle-details": VehicleDetails,
    "vehicle-location": VehicleLocation,
    "vehicle-properties": VehicleProperties,
  },
};
</script>

<style>
</style>