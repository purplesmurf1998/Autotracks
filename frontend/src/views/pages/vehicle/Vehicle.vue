<template>
  <div>
    <!-- <CButton color="secondary" @click="$router.go(-1)" class="mb-2"
      >Go back to the inventory</CButton
    > -->
    <CNav fill variant="pills" class="mb-3">
      <CNavItem :active="tab == 0" @click.native="setTab(0)">
        Vehicle
      </CNavItem>
      <CNavItem :active="tab == 1" @click.native="setTab(1)">
        History
      </CNavItem>
    </CNav>
    <hr />

    <vehicle v-if="tab == 0 && vehicle" :vehicle="vehicle" :setNewVehicle="setNewVehicle"/>
    <vehicle-history v-if="tab == 1 && vehicle" :vehicle="vehicle" />
  </div>
</template>

<script>
const axios = require("axios");

import Vehicle from "../../../modules/vehicles/Vehicle.vue";
import VehicleHistory from "../../../modules/vehicles/VehicleHistory.vue";

export default {
  name: "Vehicle",
  data() {
    return {
      tab: 0,
      vehicle: null,
    };
  },
  components: {
    vehicle: Vehicle,
    "vehicle-history": VehicleHistory,
  },
  methods: {
    setTab(tab) {
      console.log(tab);
      this.tab = tab;
    },
    setNewVehicle(newVehicle) {
      this.vehicle = newVehicle;
    },
    fetchVehicle() {
      axios({
        method: "GET",
        url: `${this.$store.state.api}/inventory/vehicle/${this.$route.params.vehicleId}`,
        headers: {
          Authorization: `Bearer ${this.$store.state.auth.token}`,
        },
      })
        .then((response) => {
          if (response.data.success) {
            this.vehicle = response.data.payload;
          }
        })
        .catch((error) => {
          this.$router.replace("/pages/404");
        });
    },
  },
  mounted() {
    this.fetchVehicle();
  },
};
</script>

<style>
</style>