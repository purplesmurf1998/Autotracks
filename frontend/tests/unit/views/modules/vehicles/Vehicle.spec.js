import {createLocalVue, shallowMount} from '@vue/test-utils'
import CoreuiVue from '@coreui/vue-pro'
import Vuex from 'vuex'

import Vehicle from './../../../../../src/modules/vehicles/Vehicle.vue'

const localVue = createLocalVue()
localVue.use(CoreuiVue)
localVue.use(Vuex)

const store = new Vuex.Store({
    state: {
      sidebarShow: 'responsive',
      sidebarMinimize: false,
      asideShow: false,
      darkMode: false
    }
  })

  describe('Vehicle.vue', () => {
    const wrapper = shallowMount(Vehicle, { 
      store, 
      localVue
    })
    it('has a name', () => {
      expect(Vehicle.name).toBe('Vehicle')
    })
    it('should render correct content', () => {
        
    })
    test('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })