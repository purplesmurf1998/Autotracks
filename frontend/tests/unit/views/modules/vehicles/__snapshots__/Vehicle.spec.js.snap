// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vehicle.vue renders correctly 1`] = `
<div>
  <crow-stub
    gutters="true"
    tag="div"
  >
    <ccol-stub
      tag="div"
    >
      <vehicle-details-stub />
    </ccol-stub>
     
    <ccol-stub
      tag="div"
    >
      <vehicle-location-stub />
    </ccol-stub>
  </crow-stub>
   
  <vehicle-properties-stub />
</div>
`;
